### PR TITLE
feat(models): Triton GraphTransformer

### PR DESCRIPTION
**Pull request by japols**
## Description
Introduce custom Triton Kernel for a fast, memory efficient and deterministic GraphTransformer.

Benchmarking results comparing the current GraphTransformer implementation (PyG) against PyG compiled and the Triton kernel on different graphs, measured in ms per fwd + bwd:

All runs use era = n320, h = lvl6-multi-scale, dtype = fp16.

A100

| Graph type |    PyG | PyG (compiled) | Triton   |
|:------------|-------:|---------------:|---------:|
| era_to_h    |  53.09 |          15.86 | **6.81** |
| h_to_h      |  20.32 |           4.45 | **1.82** |
| h_to_era    | 100.44 |          20.53 | **11.19** |
| dop_enc     |  60.44 |          17.90 | **9.08** |

H200

| Graph type |   PyG | PyG (compiled) |   Triton |
| :--------- | ----: | -------------: | -------: |
| era_to_h   | 23.97 |           6.02 | **2.77** |
| h_to_h     |  9.64 |           1.69 | **0.83** |
| h_to_era   | 46.99 |           8.16 | **4.87** |
| dop_enc    | 26.62 |           6.85 | **3.86** |

Exemplary impact on training for an o800 lvl7 GraphTransformer model in bf16 on H200s:
pyg compiled: 3.90 s/it → Triton: 2.62 s/it (×1.5 speedup)

## What problem does this change solve?
Current PyG GraphTransformer suffers from high memory footprint due to per-edge computations, multiple kernel launches and stability issues with atomics in low precision. [Compiling](https://github.com/ecmwf/anemoi-core/pull/181) can help adress some of these issues, but one can do better with a custom triton kernel.

##  Additional notes ##
Checked outputs and gradients match pyg version + small training sanity check: [mlflow](https://mlflow.ecmwf.int/#/metric?runs=[%224803881259ff447cbbb4ff4436926823%22,%22b37aee50b3664afa9fc4aee6b557d71b%22]&metric=%22train_mse_loss_step%22&experiments=[%2245%22]&plot_metric_keys=%5B%22train_mse_loss_step%22%5D&plot_layout={%22autosize%22:true,%22xaxis%22:{%22autorange%22:true,%22type%22:%22linear%22,%22range%22:[-1035.9219986020375,960.4671241229228]},%22yaxis%22:{%22range%22:[-0.6010538821019717,6.501633332403755],%22autorange%22:true,%22type%22:%22linear%22}}&x_axis=step&y_axis_scale=linear&line_smoothness=1&show_point=false&deselected_curves=[]&last_linear_y_axis_range=[])

TODO
- [x] make GT backend configurable, pyg as fallback
- [x] add pytests

***As a contributor to the Anemoi framework, please ensure that your changes include unit tests, updates to any affected dependencies and documentation, and have been tested in a parallel setting  (i.e., with multiple GPUs). As a reviewer, you are also responsible for verifying these aspects and requesting changes if they are not adequately addressed. For guidelines about those please refer to https://anemoi.readthedocs.io/en/latest/***

By opening this pull request, I affirm that all authors agree to the [Contributor License Agreement.](https://github.com/ecmwf/codex/blob/main/Legal/contributor_license_agreement.md)

### CODE REVIEW

File: models/src/anemoi/models/layers/block.py [hunk 1]
```diff
@@ -558,26 +558,19 @@ def attention_block(
         # self.conv requires size to be a tuple
         conv_size = (size, size) if isinstance(size, int) else size
 
-        if num_chunks > 1:
```
Comment (cathalobrien) [inline models/src/anemoi/models/layers/block.py:571]: Does chunking not work with the triton graph transformer?

File: models/src/anemoi/models/triton/gt.py [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+# In applying this licence, ECMWF does not waive the privileges and immunities
+# granted to it by virtue of its status as an intergovernmental organisation
+# nor does it submit to any jurisdiction.
+
+
+import torch
+import triton
+import triton.language as tl
+
+
```
Comment (cathalobrien) [inline models/src/anemoi/models/triton/gt.py:15]: Would be interesting to autotune the number of warps
```python
@triton.autotune(configs=[
    triton.Config(num_warps=2),
    triton.Config(num_warps=3),
    triton.Config(num_warps=4),
    triton.Config(num_warps=8),
  ],
key=["C"]
)
```

File: models/src/anemoi/models/layers/block.py (hunk 1 above)
Comment (japols) [reply models/src/anemoi/models/layers/block.py:571]: reintroduced it now together with the pyg gt

File: models/tests/triton/test_triton_gt.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+@pytest.mark.parametrize(
+    "n_src,n_dst,h,d,device",
+    [
+        (4, 10, 2, 4, "cuda" if torch.cuda.is_available() else "cpu"),
+    ],
+)
+def test_graph_transformer_vs_reference(n_src: int, n_dst: int, h: int, d: int, device: str):
+    """Test that triton GraphTransformerFunction matches reference implementation."""
+    edge_index, m = build_bipartite_graph(n_src, n_dst)
+    degrees = torch.bincount(edge_index[1], minlength=n_dst)
```
Comment (japols) [inline models/tests/triton/test_triton_gt.py]:
```suggestion
```

File: models/src/anemoi/models/triton/gt.py (hunk 2 above)
Comment (japols) [reply models/src/anemoi/models/triton/gt.py:15]: didn't see a big speedup from that, let's maybe do a follow up with more autotuning etc and keep it "simple" for now

File: models/src/anemoi/models/layers/block.py [hunk 4]
```diff
@@ -483,7 +486,7 @@ def __init__(
         self.lin_self = Linear(in_channels, num_heads * self.out_channels_conv, bias=bias)
         self.lin_edge = Linear(edge_dim, num_heads * self.out_channels_conv)  # , bias=False)
 
-        self.conv = GraphTransformerConv(out_channels=self.out_channels_conv)
+        # self.conv = GraphTransformerConv(out_channels=self.out_channels_conv)
```
Comment (ssmmnn11) [inline models/src/anemoi/models/layers/block.py]: remove?

File: models/src/anemoi/models/schemas/encoder.py [hunk 5]
```diff
@@ -38,7 +38,7 @@ class GraphTransformerEncoderSchema(TransformerModelComponent):
     @model_validator(mode="after")
     def check_valid_extras(self) -> Any:
         # This is a check to allow backwards compatibilty of the configs, as the extra fields are not required.
-        allowed_extras = {"shard_strategy": str}
+        allowed_extras = {"shard_strategy": str, "backend": str}
```
Comment (japols) [inline models/src/anemoi/models/schemas/encoder.py]:
```suggestion
        allowed_extras = {"shard_strategy": str, "graph_attention_backend": str}
```
